Alerting: Add details and edit pages for groups

konrad147 opened the pull request:
**What is this feature?**
This PR introduces dedicated detail and edit pages for evaluation groups. These new pages will handle all group management functionality, which is currently embedded within the alert list page.

**Why do we need this feature?**
We are moving group management actions—such as renaming, reordering, and deleting—away from the alert list page for better usability and performance. In some cases, updating a group can take a significant amount of time, and having these actions bundled into a single, dedicated space will improve the user experience. Additionally, grouping related actions (like reordering and changing intervals) in one place reduces the need for multiple modals, making the workflow more intuitive and efficient.

**Group details page**
<img width="1070" alt="image" src="https://github.com/user-attachments/assets/0d8c5b5f-d33a-4fed-a022-000369dd061e" />

**Group edit page**
<img width="1075" alt="image" src="https://github.com/user-attachments/assets/d8a76ab4-9c63-402e-aee4-75a8cb8cad28" />


**Special notes for your reviewer:**
Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

konrad147 committed:
Add basic details page for groups

konrad147 committed:
Remove unused imports

konrad147 committed:
Add basic edit page for groups

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Add functional group details page

konrad147 committed:
Improve form, add namespaces for DS groups

konrad147 committed:
Add support for multiple actions in useProduceNewRuleGroup

konrad147 committed:
Attach real actions to form submit

konrad147 committed:
Add tests for the group details page

konrad147 committed:
Add basic tests for the group edit page

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Add tests for Mimir update

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Add rule group consistency check

konrad147 committed:
Extract draggable rules table to a separate file

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Add prom consistency waiting after group saving

konrad147 commented:
/deploy-to-hg

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Add duration measure for Prometheus reconciliation time

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Remove a blinking error when redirecting to a new group

konrad147 committed:
Improve group details page. Use ruler or prom api depending on the ds capabilities

konrad147 committed:
Add group delete action for DMA

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Fix GroupDetailsPage tests

konrad147 committed:
Update tests

konrad147 committed:
Add and improve Edit page tests

konrad147 committed:
Add Group export for GMA groups

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/rules/RulesGroup.tsx:
```diff
@@ -75,6 +76,8 @@ export const RulesGroup = React.memo(({ group, namespace, expandAll, viewMode }:
     return isGrafanaRulerRule(rule.rulerRule) && rule.rulerRule.grafana_alert.provenance;
   });
 
+  const canEditGroup = hasRuler && !isProvisioned && !isFederated && canEditRules(rulesSourceName);
```
can we edit groups that contain rules provided by a plugin?

konrad147 replied:
Good point! No, we cannot - technically we can, but we shouldn't allow that in the UI

konrad147 committed:
Fix RulesGroup tests, add translations

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Disable editing plugin provided groups

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Fix alertingApi options, fix tests

konrad147 committed:
Fix lint errors, update translations

konrad147 commented:
/deploy-to-hg

gillesdemey committed:
use name for grafana managed recording rules

gillesdemey committed:
add namespace to nav

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Remove group modals from the list page

konrad147 committed:
Merge branch 'alerting/custom-groups-page' of github.com:grafana/grafana into alerting/custom-groups-page

gillesdemey committed:
add cancel button to edit form

gillesdemey committed:
add test for cancel butotn

gillesdemey committed:
fix recording rule badge for Grafana managed rules

gillesdemey reviewed (commented):
First round of review :)

konrad147 commented on public/app/features/alerting/unified/rule-editor/__snapshots__/RuleEditorCloudRules.test.tsx.snap:
```diff
@@ -2,17 +2,6 @@
 
 exports[`RuleEditor cloud can create a new cloud alert 1`] = `
 [
-  {
```
Yes, I unified parameters in the call to `useFeatureDiscovery` hook in the `useProduceNewRuleGroup` so it can used cached version of feature discovery response
https://github.com/grafana/grafana/pull/100884/files#diff-b7590878e7e4df8c718ea69cefd991818010690126cf0a09dda7604ebbab0035
The test only checks requests made after clicking submit button, so the `buildinfo` call is gone, because the value is taken from the cache

gillesdemey replied:
Awesome!

konrad147 commented on public/app/features/alerting/unified/api/alertingApi.ts:
```diff
@@ -64,6 +64,8 @@ export const backendSrvBaseQuery =
         ...requestOptions,
         ...(body && { data: body }),
+        ...(isBoolean(showSuccessAlert) && { showSuccessAlert }),
+        ...(isBoolean(showErrorAlert) && { showErrorAlert }),
```
Yes and no 😄 
We've been supporting them if passed through the `requestOptions` object, so during the endpoint configuration in RTKQ. 
We do support these indirectly if you use `successMessage` or `errorMessage` properties of `notificationOptions`, however `notificationOptions` allow you to pass also `showSuccessAlert` and `showErrorAlert` directly, and we didn't support that
There is a bit of collision because we can get these parameters from both `requestOptions` and `notificationOptions`

konrad147 commented on public/app/features/alerting/unified/group-details/GroupDetailsPage.tsx:
```diff
@@ -0,0 +1,306 @@
+    rules: group.rules.map<RuleDetails>((rule) => {
+      const name = getRuleName(rule);
+
+      if (isAlertingRulerRule(rule)) {
```
Yes! good catch

konrad147 committed:
Add doc comments, improve code

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented:
/deploy-to-hg

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented:
/deploy-to-hg

konrad147 commented:
/deploy-to-hg

konrad147 committed:
Move url changes to be the last action in form submit

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 reviewed (commented):
/deploy-to-hg

konrad147 commented:
/deploy-to-hg

gillesdemey commented on public/app/features/alerting/unified/hooks/usePrometheusConsistencyCheck.ts:
```diff
@@ -43,6 +52,172 @@ function useMatchingPromRuleExists() {
+  const [fetchPrometheusRuleGroups] = useLazyPrometheusRuleNamespacesQuery();
+  const [fetchRuleGroup] = useLazyGetRuleGroupForNamespaceQuery();
+
+  const isGroupInSync = useCallback(
```
Shouldn't we use `useAsync` here so we can track progress for this check somehow?

gillesdemey commented on public/app/features/alerting/unified/hooks/usePrometheusConsistencyCheck.ts:
```diff
@@ -43,6 +52,172 @@ function useMatchingPromRuleExists() {
+  return { isGroupInSync };
+}
+
+export function useRuleGroupConsistencyCheck() {
```
Discussed offline, but perhaps worth looking in to a similar pattern I've used in the AM settings page for waiting on remote state by invalidating cache until some condition was met;

https://github.com/grafana/grafana/blob/07b069db83b8a2415a7e2125736aadbf3b6cc1d2/public/app/features/alerting/unified/components/settings/SettingsContext.tsx#L172-L187

gillesdemey replied:
not a blocker; so I'm marking this as resolved ;)

konrad147 commented on public/app/features/alerting/unified/group-details/GroupEditPage.tsx:
```diff
@@ -0,0 +1,358 @@
+  } = useDiscoverDsFeaturesQuery({ uid: ruleSourceUid });
+
+  // We use useAsync instead of RTKQ query to avoid cache invalidation issues when the group is being deleted
+  // RTKQ query would refetch the group after it's deleted and we'd end up with a blinking group not found error
```
In this case, the main reason is that we have no stable identifiers for groups, but we need to use names as identifiers. Changing the name, changes also the URL which triggers refetching and so on

konrad147 commented on public/app/features/alerting/unified/hooks/ruleGroup/useUpdateRuleGroup.ts:
```diff
@@ -16,6 +25,120 @@ import { useProduceNewRuleGroup } from './useProduceNewRuleGroup';
+  ruleSwaps?: SwapOperation[];
+}
+
+export function useUpdateRuleGroup() {
```
I will, and I will deprecate the old ones, wdyt?

tomratcliffe commented:
Some initial thoughts on the UX (sorry this has come later on after you having worked on it for a while, but there's not anything major!):

* Do we need a separate view/edit page? Could we do a similar pattern to elsewhere where we have the view page just being a disabled version of the edit one?
* There isn't an easy way to go from Grouped view -> seeing the Grafana folder. Previously you could open the modal and then click the folder icon, but in the grouped view you can't do this of course, and the new view/edit pages don't include a way to get to the folder page. Thinking of the use case of wanting to edit permissions etc
* Could we put a back button on the view/edit page, so you can easily return to the list page? You can't go back to the full list view without using browser back button
* Similarly, if you go `edit` -> `cancel`, it always takes you back to the `view` page, even if you came straight from the list view
* Should the edit page form elements have a shorter width? With them at 100% they look a bit... lost?

gillesdemey commented:
> Do we need a separate view/edit page?

I personally prefer this from a technical perspective because it simplifies things around access control, on the UX side we get a few nice things like being able to use the page metadata and such (though we could also use that in the edit view).

> the new view/edit pages don't include a way to get to the folder page

Good point! We could probably add a link here for Grafana-managed rule groups in the page meta.

> Could we put a back button on the view/edit page? /  edit -> cancel, it always takes you back to the view page

I think it'd be nice to have that button + have it respect any potential `?returnTo` param. Alternatively we can use the `WithReturnButton` component which uses the `returnToPrevious()` hook.

> Should the edit page form elements have a shorter width?

Maybe we should defer on this for now, don't though I tend to agree.

gillesdemey committed:
Add returnTo URL handling for alert rule group navigation

gillesdemey committed:
Create dedicated Title component showing breadcrumb navigation between folder
and group name.

Add label distinction between folders and namespaces based on
the rule source (Grafana vs external).

gillesdemey commented:
Addressed some UX concerns in two commits

konrad147 commented:
> Do we need a separate view/edit page? Could we do a similar pattern to elsewhere where we have the view page just being a disabled version of the edit one?

It could make the Edit page much more complex. Edit page users the Ruler endpoint exclusively, so it works only with Ruler DTOs. On the other hand, the view page needs to also work for vanilla prometheus data sources, which means it needs to work with both Ruler and Prometheus DTOs. There is a bit of glue code to make it work for both on the view page

Initially I wanted to use only Prom endpoint for the view page, but to provide more consistency we decided to use Ruler if available

> Could we put a back button on the view/edit page, so you can easily return to the list page? You can't go back to the full list view without using browser back button

I have no strong opinion about this. I haven't seen this patter anywhere besides Alerting, so I haven't added it

Regarding the `returnTo` parameter, I'd love to remove it from Alerting in the future, as using it poses a lot of challenges, so intentionally didn't implement it here

konrad147 committed:
Address PR feedback, minor refactorings

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Update rule group links to include return path and refactor rule type checks

- Modified `RulesGroup` and `GroupDetailsPage` components to include `includeReturnTo` in edit page links.
- Refactored rule type checks in `DraggableRulesTable` and `GroupDetailsPage` to use `rulerRuleType` for better clarity and maintainability.
- Updated documentation in `useUpdateRuleGroup` to clarify functionality for updating or moving rule groups.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 commented on public/app/features/alerting/unified/group-details/components/DraggableRulesTable.tsx:
```diff
@@ -0,0 +1,188 @@
+  const numberEvaluationsToStartAlerting = getNumberEvaluationsToStartAlerting(pendingPeriod ?? '0s', groupInterval);
+  const isRecordingRule = isGrafanaOrDataSourceRecordingRule(rule);
+
+  // TODO Bring back isClone and isDraggin styles
```
Done!

konrad147 committed:
Refactor RulesGroup component and tests for improved link handling and permissions checks

- Added `includeReturnTo` parameter to rule group detail links in `RulesGroup` for better navigation.
- Updated test cases to verify rendering of edit and view buttons based on user permissions.
- Simplified test setup by removing unnecessary Redux provider wrapping in tests.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Refactor: Update routing and test assertions in GroupDetails and GroupEdit pages

- Modified route paths in GroupDetailsPage and GroupEditPage tests to use `dataSourceUid` instead of `sourceId`.
- Updated test assertions to reflect changes in folder title and link structure in GroupDetailsPage.
- Simplified Title component by removing folder-related props and logic, focusing solely on the group name.

konrad147 commented:
/deploy-to-hg

soniaAguilarPeiron reviewed (approved):
LGTM! 🚀 great work!

konrad147 committed:
Refactor: Simplify Title rendering in GroupDetailsPage

- Updated the renderTitle function in GroupDetailsPage to remove the folder prop from the Title component, focusing solely on the group name.

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Update GroupDetailsPage to prevent editing of provisioned groups

konrad147 committed:
Fix imports

konrad147 committed:
Improve styles

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Fix navigation when served from subpath

konrad147 committed:
Merge branch 'main' of github.com:grafana/grafana into alerting/custom-groups-page

konrad147 committed:
Improve group removal handling in Prom consistency check. Fix Delete group button

gillesdemey reviewed (approved):
LGTM ✅ 🚢
